client-server-client/src/main/scala/zio/raft/client/RaftClient.scala

```diff
@@ -0,0 +1,340 @@
+package zio.raft.client
+
+import zio._
+import zio.stream._
+import zio.raft.protocol._
+import scodec.bits.ByteVector
+import java.time.Instant
+
+/**
+ * Main RaftClient implementation using pure functional state machine.
+ * 
+ * The client uses a unified stream that merges all events.
+ * Each ClientState knows how to handle events and transition to new states.
+ */
+class RaftClient private (
+  zmqTransport: ZmqClientTransport,
+  config: ClientConfig,
+  actionQueue: Queue[ClientAction]
+) {
+  
+  def submitCommand(payload: ByteVector): Task[ByteVector] = 
+    for {
+      promise <- Promise.make[Throwable, ByteVector]
+      action = ClientAction.SubmitCommand(payload, promise)
+      _ <- actionQueue.offer(action)
+      result <- promise.await
+    } yield result
+  
+  def disconnect(): UIO[Unit] = 
+    actionQueue.offer(ClientAction.Disconnect).unit
+}
+
+object RaftClient {
+  
+  def make(addresses: List[String], capabilities: Map[String, String]): ZIO[Scope, Throwable, RaftClient] = 
+    for {
+      config = ClientConfig.make(addresses, capabilities)
+      validatedConfig <- ClientConfig.validated(config).mapError(new IllegalArgumentException(_))
+      
+      zmqTransport <- createZmqTransport(validatedConfig)
+      actionQueue <- Queue.unbounded[ClientAction]
+      
+      client = new RaftClient(zmqTransport, validatedConfig, actionQueue)
+      
+      _ <- startMainLoop(client, zmqTransport, validatedConfig, actionQueue).fork
+      
+      // Send initial CreateSession
+      nonce <- Nonce.generate()
+      _ <- zmqTransport.sendMessage(CreateSession(validatedConfig.capabilities, nonce))
+      
+    } yield client
+  
+  private def createZmqTransport(config: ClientConfig): UIO[ZmqClientTransport] = 
+    ZIO.succeed(new ZmqClientTransportStub())
+  
+  private def startMainLoop(
+    client: RaftClient,
+    transport: ZmqClientTransport,
+    config: ClientConfig,
+    actionQueue: Queue[ClientAction]
+  ): Task[Unit] = {
+    
+    val actionStream = ZStream.fromQueue(actionQueue)
+      .map(StreamEvent.Action(_))
+    
+    val messageStream = transport.incomingMessages
+      .map(StreamEvent.ServerMsg(_))
+    
+    val keepAliveStream = ZStream.tick(config.keepAliveInterval)
+      .map(_ => StreamEvent.KeepAliveTick)
+    
+    val unifiedStream = actionStream
+      .merge(messageStream)
+      .merge(keepAliveStream)
+    
+    val initialState: ClientState = ClientState.Connecting(
+      capabilities = config.capabilities,
+      nonce = None,
+      addresses = config.clusterAddresses,
+      currentAddressIndex = 0,
+      retryManager = RetryManager.empty(config)
+    )
+    
+    // Pure functional state machine: state handles events
+    unifiedStream
+      .runFold(initialState) { (state, event) =>
+        state.handle(event, transport, config, client)
+      }
+      .unit
+  }
+  
+  private def generateRequestId(): UIO[RequestId] = 
```

Comment:
request id should be sequential, don't get a random one. Use a Ref[RequestId] within each State, it should pass between state and zeroed out on Disconnected (together with SessionId)